rstudio/rstudio
# Implement code diagnostics in Visual Mode

### Intent

This change adds support for code diagnostics inside code chunks in Visual Mode. 

![image](https://user-images.githubusercontent.com/470418/143282181-3a20e00e-5a90-49ae-9656-d8e4493bcbb9.png)

Addresses https://github.com/rstudio/rstudio/issues/9874.

### Approach

Enabling diagnostics is not that difficult; the hard part is making it sufficiently performant. The existing implementation requires that the document be saved, and the linter runs against the saved copy of the document from the source database. That's a little slow when there's an idle timer and Pandoc process in front of the save. So:

- The diagnostics engine has been modified so that it can run on a code fragment in addition to an entire document.
- The `LintManager`'s dependency on `TextEditingTarget` has been broken; it now takes a `LintSource` that can be either a source editor or a visual chunk.
- The visual chunks can get lint from both the parent editor (when diagnostics for the whole document are retrieved) and from their own embedded `LintManager`. 
- Visual chunks are re-linted when their contents change (after an idle period); this re-lint uses only the content of the chunk (delivered from the front end instead of read from the source database) for performance reasons enumerated above.

The other big part of this change is managing the lint symbols in the "gutter" area to the left of the chunk. We have to manage and position these ourselves, unless the line numbers are turned on for chunks (in which case Ace has a gutter area we can use). 

### Automated Tests

See https://github.com/rstudio/rstudio-ide-automation/issues/106.

### QA Notes

- Very broadly, source diagnostics should now work in visual mode very similarly to source mode.
- We are slightly less aggressive about triggering a scan for problems; editing a chunk (making a change to it) should trigger a diagnostics scan if you're not sure whether one has been run.
- A couple of optional diagnostics settings do not work in the visual editor; specifically, the warnings about defined but not used / used but not defined symbols are not shown. 
- Look out for problems drawing the diagnostics warnings on the correct line, and/or showing the gutter warnings in the right place.
- The "show line numbers in the visual editor" setting affects how we draw warnings as noted above; try testing with it off and on. 

### Checklist
